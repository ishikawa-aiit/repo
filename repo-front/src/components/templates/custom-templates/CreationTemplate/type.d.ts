import { CustomTemplateTypes, MetaType } from '@/components/types'

export namespace CustomTemplateType {
  export type Props = {
    customTemplates: CustomTemplateTypes.CustomTemplateProps[]
  } & MetaType.Props

  export type TemplateFormElementsPropType = {
    /** 項目名 */
    name: keyof Inputs
    /** 表示用の項目名 */
    label: string
    /** プレースホルダー */
    placeholder?: string
  }

  export type ModelSentencePropsType = {
    /** 項目名 */
    name: keyof Inputs
    /** 例文 */
    modelSentence: string
  }

  export type ExplainSentencePropsType = {
    /** 項目名 */
    name: string
    /** 例文 */
    explainSentence: string
  }

  export type Template = {
    /** 項目名 */
    name: keyof Inputs
    /** テンプレートの内容 */
    sentence: string
  }

  export type Inputs = {
    /** テンプレートID */
    id: number | null
    /** テンプレート名 */
    name: string
    /** 活動時間 */
    activityTime?: string
    /** これまでの活動と成果の実績 */
    doneActivity?: string
    /** 今後の活動と成果の予定 */
    todoActivity?: string
    /** 課題と解決策 */
    solution?: string
    /** できごと・気づき */
    event?: string
    /** 特記事項 */
    remark?: string
    /** PTで共有する */
    isShared: boolean
  }
}
