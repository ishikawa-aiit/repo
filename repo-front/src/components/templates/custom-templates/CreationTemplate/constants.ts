import { CustomTemplateType } from './type'

import { CustomTemplateTypes } from '@/components/types'

/**
 * テンプレートのInputコンポーネントを作成する時用
 */
export const templateInputProps: CustomTemplateType.TemplateFormElementsPropType[] = [
  {
    name: 'name',
    label: 'テンプレート名',
    placeholder: 'テンプレートの名前を入力してください',
  },
]

/**
 * テンプレートのTextareaコンポーネントを作成する時用
 */
export const templateTextareaProps: CustomTemplateType.TemplateFormElementsPropType[] = [
  {
    name: 'activityTime',
    label: '活動時間',
    placeholder: '今週の活動とそれにかかった時間を記述してください',
  },
  {
    name: 'doneActivity',
    label: '今週の活動と成果の実績',
    placeholder: '今週のチーム及び個人の活動と、その結果の成果物の状況を記述してください',
  },
  {
    name: 'todoActivity',
    label: '来週の活動と成果の予定',
    placeholder: '次週のチーム及び個人の活動の目標と期待される成果物について、計画を記述してください',
  },
  {
    name: 'solution',
    label: '課題と解決策',
    placeholder: '今週発生した課題と、今週解決した課題及びその解決策を記述してください',
  },
  {
    name: 'event',
    label: 'できごと・気づき',
    placeholder: 'プロジェクト全体で、印象に残る出来事や自分の気づきについて記述してください',
  },
  {
    name: 'remark',
    label: '特記事項',
    placeholder: '他の項目での記述内容以外で報告したい事柄等について記述してください',
  },
]

/**
 * 例文を作成する時用
 */
export const modelSentences: CustomTemplateType.ModelSentencePropsType[] = [
  {
    name: 'name',
    modelSentence: 'チームと個人テンプレート',
  },
  {
    name: 'activityTime',
    modelSentence: 'チーム作業: ○時間○分\n個人作業: ○時間○分',
  },
  {
    name: 'doneActivity',
    modelSentence: 'チームの活動と成果の実績\n・\n・\n・\n\n個人の活動と成果の実績\n・\n・\n・',
  },
  {
    name: 'todoActivity',
    modelSentence: 'チームの活動と成果の予定\n・\n・\n・\n\n個人の活動と成果の予定\n・\n・\n・',
  },
  {
    name: 'solution',
    modelSentence: 'チームの課題と解決策\n課題: \n解決策: \n\n個人の課題と解決策\n課題: \n解決策: ',
  },
  {
    name: 'event',
    modelSentence: 'チームのできごと・気づき\n・\n・\n\n個人のできごと・気づき\n・\n・',
  },
  {
    name: 'remark',
    modelSentence: 'チームの特記事項\n・\n\n個人の特記事項\n・',
  },
]

/**
 * テンプレートのSelectコンポーネントを作成する時用
 */
export const customTemplateSelectProp: CustomTemplateType.TemplateFormElementsPropType = {
  name: 'id',
  label: '編集対象のテンプレート',
}

/**
 * 説明文を作成する時用
 */
export const explainSentences: CustomTemplateType.ExplainSentencePropsType[] = [
  {
    name: 'selectTemplate',
    explainSentence:
      '「新規作成」を選択すると、テンプレートを新規作成することができます。テンプレート名を選択すると、テンプレートを編集することができます。テンプレートを選択すると内容をフォームに反映します。 自身が作成した週報のみ編集可能です。',
  },
  {
    name: 'shareSwitch',
    explainSentence: '',
  },
]

/**
 * 新規作成用テンプレート
 */
export const templateForNewCreate: CustomTemplateTypes.CustomTemplateProps = {
  id: null,
  name: '新規作成',
  activityTime: '',
  doneActivity: '',
  todoActivity: '',
  solution: '',
  event: '',
  remark: '',
  team: null,
}

/**
 * PT共有フラグコンポーネントを作成するとき用
 * teamなしで不活性化するため単独で作成
 */
export const isSharedProp: CustomTemplateType.TemplateFormElementsPropType = { name: 'isShared', label: 'PTで共有する' }

/**
 * INPUTとTEXTAREAの最大入力文字数
 */
export const templateCharacterLimitConst = {
  MAX_INPUT: 250,
  MAX_TEXTAREA: 16000,
} as const

/**
 * バリデーションエラーメッセージ
 */
type validationType = 'MAX' | 'REQUIRE'

export const getValidationErrorMessage = (type: validationType, numOfChars = 0) => {
  switch (type) {
    case 'REQUIRE':
      return '必ず入力してください'
    case 'MAX':
      return `${numOfChars}文字以内で入力してください`
  }
}
