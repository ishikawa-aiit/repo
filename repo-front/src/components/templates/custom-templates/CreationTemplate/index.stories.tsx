import { ComponentMeta, ComponentStoryObj } from '@storybook/react'

import { CreationTemplate } from '.'

export default {
  component: CreationTemplate,
} as ComponentMeta<typeof CreationTemplate>

const sampleData = [
  {
    id: 1,
    name: 'dummy1',
    activityTime: 'dummy1',
    doneActivity: 'dummy1',
    todoActivity: 'dummy1',
    solution: 'dummy1',
    event: 'dummy1',
    remark: 'dummy1',
    team: {
      id: 1,
      name: 'dummy1',
      createdAt: 'dummy1',
      updatedAt: 'dummy1',
    },
  },
  {
    id: 2,
    name: 'dummy2',
    activityTime: 'dummy2',
    doneActivity: 'dummy2',
    todoActivity: 'dummy2',
    solution: 'dummy2',
    event: 'dummy2',
    remark: 'dummy2',
    team: {
      id: 2,
      name: 'dummy2',
      createdAt: 'dummy2',
      updatedAt: 'dummy2',
    },
  },
]

export const CreationTemplateUsage: ComponentStoryObj<typeof CreationTemplate> = {
  storyName: '仕様',
  argTypes: {
    customTemplates: {
      defaultValue: sampleData,
    },
  },
  parameters: {
    layout: 'fullscreen',
    docs: {
      description: {
        component: 'テンプレート作成ページ（"/create-templates/create"）用のテンプレートとなるコンポーネントです',
      },
    },
  },
}
