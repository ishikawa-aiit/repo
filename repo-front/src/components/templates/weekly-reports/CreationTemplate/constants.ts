import { CreationTemplateType } from './type'

/**
 * 週報のSelectコンポーネントを作成する時用
 */
export const weeklyReportSelectProp: CreationTemplateType.FormElementsPropType[] = [
  {
    name: 'template',
    label: 'テンプレート',
    placeholder: '選択してください',
  },
  {
    name: 'requirementId',
    label: '報告週',
    placeholder: '選択してください',
  },
]

/**
 * 週報のTextareaコンポーネントを作成する時用
 */
export const weeklyReportTextareaProps: CreationTemplateType.FormElementsPropType[] = [
  {
    name: 'activityTime',
    label: '活動時間',
    placeholder: '今週の活動とそれにかかった時間を記述してください',
  },
  {
    name: 'doneActivity',
    label: '今週の活動と成果の実績',
    placeholder: '今週のチーム及び個人の活動と、その結果の成果物の状況を記述してください',
  },
  {
    name: 'todoActivity',
    label: '来週の活動と成果の予定',
    placeholder: '次週のチーム及び個人の活動の目標と期待される成果物について、計画を記述してください',
  },
  {
    name: 'solution',
    label: '課題と解決策',
    placeholder: '今週発生した課題と、今週解決した課題及びその解決策を記述してください',
  },
  {
    name: 'event',
    label: 'できごと・気づき',
    placeholder: 'プロジェクト全体で、印象に残る出来事や自分の気づきについて記述してください',
  },
  {
    name: 'remark',
    label: '特記事項',
    placeholder: '他の項目での記述内容以外で報告したい事柄等について記述してください',
  },
]

/**
 * 例文を作成する時用
 */
export const modelSentences: CreationTemplateType.ModelSentencePropsType[] = [
  {
    name: 'activityTime',
    modelSentence: 'チーム作業: 9時間30分\n個人作業: 9時間0分',
  },
  {
    name: 'doneActivity',
    modelSentence:
      'チームの活動と成果の実績\n・モブプログラミングを用いて、チームで一覧ページを実装した。\n・来週のマンスリーレビューの準備のため、マンスリーレビューのアジェンダを相談して決定した。\n・書籍『テスト駆動開発』の輪読会を実施し、テスト駆動開発の方法を学び、チームで実践する方法を話し合った。\n\n個人の活動と成果の実績\n・トップページを実装し、他メンバーにレビューを依頼した。\n・来週の準備として、着手しそうなタスクの細分化を行った。\n・書籍『React実践の教科書』の「再レンダリングの仕組みと最適化」を読み、メモ化の必要性とその方法について学んだ。',
  },
  {
    name: 'todoActivity',
    modelSentence:
      'チームの活動と成果の予定\n・モブプログラミングを用いて、チームでログインページを実装する。\n・マンスリーレビューがあるため、リハーサルを行う。\n・テスト駆動開発の実践について、具体的な計画をたてる。\n\n個人の活動と成果の予定\n・他メンバーからのレビューをもとに、トップページの品質をブラッシュアップする。\n・細分化したタスクに対して見積もりを行い、担当者を割り振る。\n・書籍『React実践の教科書』の「グローバルなState 管理」を読む。',
  },
  {
    name: 'solution',
    modelSentence:
      'チームの課題と解決策\n課題: チームでモブプログラミングを初めて実施したが、モブプログラミングの方法についてメンバー間で認識の齟齬があり、途中で混乱してしまった。\n解決策: 新しい方法を試す際には、実施前に方法や目的の認識合わせを行うこととする。\n\n個人の課題と解決策\n課題: トップページのレビューを依頼した際に、全体的なレイアウトに関する指摘を受け、修正対応が多くなってしまった。\n解決策: コードを書く前に、全体的なレイアウトや実装の方針についてチームに確認することで、修正対応の量を減らす。',
  },
  {
    name: 'event',
    modelSentence:
      'チームのできごと・気づき\n・モブプログラミングを初めて実施したが、ひとつのことに悩む時間が短くなり、実装時間の短縮に繋がることを実感した。また、他メンバーの目線を入れられるので、品質が高くなることを実感した。\n・ミーティングを実施する際に、最初にアイスブレイクを挟むメンバーがいた。これにより、普段あまり発言をしないメンバーも発言数が増えていたため、自分もアイスブレイクを行っていきたい。\n\n個人のできごと・気づき\n・コードレビューを通してセキュリティ観点からのコメントをもらったが、これは自分はまったく思いつかなかった。今後コードを書く際にセキュリティについても気をつけていきたい。\n・タスクを細分化するにあたり、プロダクト全体について考える必要があった。これを通して、普段より広い視点でプロダクトを俯瞰できるようになった。',
  },
  {
    name: 'remark',
    modelSentence: 'チームの特記事項\nなし\n\n個人の特記事項\nなし',
  },
]

const listTemplate = new Array(3).fill('・').join('\n')

/**
 * チームと個人（デフォルト）
 */
export const templateDefaultExists = {
  id: null,
  name: 'チームと個人',
  activityTime: 'チーム作業: ○時間○分\n個人作業: ○時間○分',
  doneActivity: `チームの活動と成果の実績\n${listTemplate}\n\n個人の活動と成果の実績\n${listTemplate}`,
  todoActivity: `チームの活動と成果の予定\n${listTemplate}\n\n個人の活動と成果の予定\n${listTemplate}`,
  solution: 'チームの課題と解決策\n課題: \n解決策: \n\n個人の課題と解決策\n課題: \n解決策: ',
  event: `チームのできごと・気づき\n${listTemplate}\n\n個人のできごと・気づき\n${listTemplate}`,
  remark: 'チームの特記事項\nなし\n\n個人の特記事項\nなし',
  team: null,
}
