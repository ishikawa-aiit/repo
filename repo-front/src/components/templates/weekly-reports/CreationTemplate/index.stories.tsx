import { ComponentMeta, ComponentStoryObj } from '@storybook/react'

import { CreationTemplate } from '.'

export default {
  component: CreationTemplate,
} as ComponentMeta<typeof CreationTemplate>

export const CreationTemplateUsage: ComponentStoryObj<typeof CreationTemplate> = {
  storyName: '仕様',
  argTypes: {},
  parameters: {
    layout: 'fullscreen',
    docs: {
      description: {
        component: 'ホームページ（"/"）用のテンプレートとなるコンポーネントです',
      },
    },
  },
}
