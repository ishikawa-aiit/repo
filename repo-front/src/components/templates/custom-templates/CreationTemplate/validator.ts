import * as Yup from 'yup'

import { templateCharacterLimitConst, getValidationErrorMessage } from './constants'
import { CustomTemplateType } from './type'

const { MAX_INPUT, MAX_TEXTAREA } = templateCharacterLimitConst

// バリデーションエラーメッセージ定義
Yup.setLocale({
  mixed: {
    required: getValidationErrorMessage('REQUIRE'),
  },
  string: {
    max: ({ max }) => {
      return getValidationErrorMessage('MAX', max)
    },
  },
})

/**
 * カスタムテンプレートページのバリデーションスキーマ
 */
export const templateValidationSchema: Yup.SchemaOf<Omit<CustomTemplateType.Inputs, 'id' | 'isShared'>> =
  Yup.object().shape({
    name: Yup.string().trim().max(MAX_INPUT).required(),
    activityTime: Yup.string().max(MAX_TEXTAREA),
    doneActivity: Yup.string().max(MAX_TEXTAREA),
    todoActivity: Yup.string().max(MAX_TEXTAREA),
    solution: Yup.string().max(MAX_TEXTAREA),
    event: Yup.string().max(MAX_TEXTAREA),
    remark: Yup.string().max(MAX_TEXTAREA),
    isShared: Yup.boolean(),
  })
