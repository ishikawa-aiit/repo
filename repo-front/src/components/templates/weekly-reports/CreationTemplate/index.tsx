import { useDisclosure } from '@chakra-ui/hooks'
import {
  Box,
  VStack,
  Center,
  Select,
  HStack,
  FormControl as ChakraFormControl,
  FormLabel,
  Flex,
} from '@chakra-ui/react'
import { Text } from '@chakra-ui/react'
import { DateTime as DateTime } from 'luxon'
import { useRouter } from 'next/router'
import { FC, memo, useEffect, useRef } from 'react'
import { useForm } from 'react-hook-form'
import { useQueryClient } from 'react-query'

import { weeklyReportSelectProp, weeklyReportTextareaProps, modelSentences } from './constants'
import { CreationTemplateType } from './type'

import { Button, FormControl, Textarea, Link, showToastOutOfService } from '@/components/molecules'
import { DefaultLayout, ModelSentenceLayout, TemplateModal } from '@/components/organisms'
import { RequirementTypes, WeeklyReportsTypes } from '@/components/types'
import { usePostWeeklyReport } from '@/hooks/api'
import { QueryKeys } from '@/types'
import {
  convertTargetDurationStartEndToReportWeekString,
  isInSubmissionDuration,
  sortWeeklyReportQueryData,
} from '@/utils'

export const CreationTemplate: FC<CreationTemplateType.Props> = memo(({ customTemplates, requirements, ...props }) => {
  const { mutate } = usePostWeeklyReport()
  const router = useRouter()
  const { isOpen, onOpen, onClose } = useDisclosure()
  const templateSelectRef = useRef<HTMLSelectElement>(null)
  const { register, handleSubmit, setValue, getValues, watch } = useForm<CreationTemplateType.Inputs>()
  const queryClient = useQueryClient()
  const templateSelectProp = weeklyReportSelectProp[0]
  const weekSelectProp = weeklyReportSelectProp[1]

  // 初回のテンプレート反映時、フォームの値をセットしても、テキストエリアが値が入力されたことを認識しない
  // そのため、フォームの値をwatch（https://react-hook-form.com/api/useform/watch/）で常時監視し、
  // 変更が加わり次第、レンダリングを行わせる
  useEffect(() => {
    watch()
  }, [watch])

  const applyTemplate = () => {
    customTemplates.forEach((customTemplate, index) => {
      if (index === templateSelectRef?.current?.selectedIndex) {
        ;(Object.keys(customTemplate) as (keyof CreationTemplateType.Inputs)[]).forEach(async (key) => {
          if (['id', 'name', 'template', 'team'].includes(key)) {
            return
          }
          setValue(key, (customTemplate[key] as unknown as string) ?? '')
        })
      }
    })
  }

  /**
   * 報告週セレクトボックスで使用するリストを作成する
   */
  const requirementDisplays: CreationTemplateType.RequirementDisplayType[] = requirements.map((requirement) => {
    return {
      id: requirement.id === null || requirement.id === undefined ? '' : requirement.id.toString(),
      name: convertTargetDurationStartEndToReportWeekString(
        requirement.targetDuration.startAt,
        requirement.targetDuration.endAt,
      ),
    }
  })

  /**
   * 初期表示する報告要件IDをstring型で返す、初期表示対象の報告要件IDがない場合はundefinedを返す
   *
   * @returns 報告要件ID
   */
  const getSelectedRequirementId = () => {
    // 現在時刻
    const dateNow: DateTime = DateTime.now()
    // 現在時刻が受付期限内となる報告要件のリストを取得する
    const isInSubmissionDurationRequirements: RequirementTypes.RequirementProps[] = requirements.filter(
      (requirement: RequirementTypes.RequirementProps) => {
        return isInSubmissionDuration(
          requirement.submissionDuration.startAt,
          requirement.submissionDuration.endAt,
          dateNow,
        )
      },
    )
    // 現在時刻が受付期限内となる報告要件が存在しない場合はundefinedとする
    if (isInSubmissionDurationRequirements.length === 0) {
      return undefined
    }
    // より受付終了日時の早い報告要件の報告要件IDを返す
    return isInSubmissionDurationRequirements
      .reduce((req, reqCompare) => {
        return req.submissionDuration.endAt < reqCompare.submissionDuration.endAt ? req : reqCompare
      })
      .id.toString()
  }

  const onSubmit = async (values: CreationTemplateType.Inputs) => {
    const requirement = requirements.find((item) => {
      return item.id.toString() === values.requirementId
    })

    if (requirement === undefined) {
      console.error('Requirement is not found.')
      showToastOutOfService()
      return
    }

    await queryClient.cancelQueries(QueryKeys.GET_WEEKLY_REPORTS)
    // 作成に失敗した場合に戻すデータは現在登録されてるデータ
    const rollbackQueryData =
      queryClient.getQueryData<WeeklyReportsTypes.ReportsProps>(QueryKeys.GET_WEEKLY_REPORTS) || []
    // 楽観的更新で使用するデータは元のデータと作成したデータをソートしたもの
    const temporaryQueryData = sortWeeklyReportQueryData([
      { requirement, ...values },
      ...rollbackQueryData,
    ]) as WeeklyReportsTypes.ReportsProps

    // 次ページで表示する週報
    const hash = temporaryQueryData.findIndex((item) => {
      // 作成前のデータは id が存在しない
      return item.id === undefined
    })
    // findIndex は要素が見つからない場合は -1 を返す
    if (hash === -1) {
      console.error('Hash is not found.')
      return
    }

    mutate({ request: values, temporaryQueryData, rollbackQueryData })

    router.push(`/weekly-reports/list#${hash}`)
  }

  const onClickApplyTemplateButton = () => {
    const { activityTime, doneActivity, todoActivity, solution, event, remark } = getValues()
    const templateValue = {
      activityTime,
      doneActivity,
      todoActivity,
      solution,
      event,
      remark,
    }

    const hasFilledField = (
      Object.keys(templateValue) as (keyof Omit<CreationTemplateType.Inputs, 'template' | 'name' | 'requirementId'>)[]
    ).some((key) => {
      return !!templateValue[key]
    })

    if (hasFilledField) {
      onOpen()
      return
    }

    applyTemplate()
  }

  return (
    <DefaultLayout heading="週報を作成する" innerWidth="100%" {...props}>
      <form onSubmit={handleSubmit(onSubmit)}>
        <ModelSentenceLayout modelSentences={modelSentences}>
          <Box gridArea="top_left">
            <VStack spacing={8}>
              <HStack maxW="100%" alignSelf="flex-start" mb={6} spacing={4}>
                <ChakraFormControl>
                  <Flex>
                    <Flex alignItems="center" w={28}>
                      <FormLabel m={0} fontWeight="normal">
                        {templateSelectProp.label}
                      </FormLabel>
                    </Flex>
                    <Box flex={1}>
                      <Select
                        data-cy={`Select_${templateSelectProp.name}`}
                        data-testid={`Select_${templateSelectProp.name}`}
                        {...register(templateSelectProp.name)}
                        ref={templateSelectRef}
                      >
                        {customTemplates.map(({ name }, index) => {
                          return (
                            <option value={name} key={index}>
                              {name}
                            </option>
                          )
                        })}
                      </Select>
                    </Box>
                  </Flex>
                </ChakraFormControl>
                <Button
                  data-testid="ApplyTemplateButton"
                  data-cy="ApplyTemplateButton"
                  type="button"
                  onClick={onClickApplyTemplateButton}
                >
                  反映する
                </Button>
              </HStack>
            </VStack>

            <Box marginBottom={20}>
              <Text data-testid="NaviLinkBeforeText" as="small">
                「
                <Link href="/custom-templates/create" text="テンプレートを作成・編集する" />
                」からテンプレートを作成することができます。
                <br />
                テンプレートを作成することで、週報を作成する際にテンプレートの内容から週報を書き始めることができます。
                先生から週報の書き方に指定があった際に、そのフォーマットをテンプレートとして登録しておくと便利です。
              </Text>
            </Box>

            <VStack spacing={8}>
              <FormControl label={weekSelectProp.label}>
                <Select
                  data-testid={`Select_${weekSelectProp.name}`}
                  placeholder={weekSelectProp.placeholder}
                  {...register(weekSelectProp.name)}
                  defaultValue={getSelectedRequirementId()}
                >
                  {requirementDisplays.map((requirementDisplay) => {
                    return (
                      <option value={requirementDisplay.id} key={requirementDisplay.id}>
                        {requirementDisplay.name}
                      </option>
                    )
                  })}
                </Select>
              </FormControl>
            </VStack>
          </Box>

          {weeklyReportTextareaProps.map((textareaProp, index) => {
            return (
              <Box gridArea={`${textareaProp.name}_left`} key={index}>
                <FormControl label={textareaProp.label}>
                  <Textarea
                    placeholder={textareaProp.placeholder}
                    {...register(textareaProp.name)}
                    data-testid={`TextArea_${textareaProp.name}`}
                  />
                </FormControl>
              </Box>
            )
          })}

          <Box gridArea="button_left">
            <Center>
              <Button type="submit">保存する</Button>
            </Center>
          </Box>
        </ModelSentenceLayout>
      </form>
      <TemplateModal isOpen={isOpen} onClose={onClose} onApply={applyTemplate} />
    </DefaultLayout>
  )
})
