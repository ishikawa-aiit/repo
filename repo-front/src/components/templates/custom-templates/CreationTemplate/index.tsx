import { Box, Center, Input, Select } from '@chakra-ui/react'
import { yupResolver } from '@hookform/resolvers/yup'
import { FC, memo, useState, useEffect } from 'react'
import { useForm } from 'react-hook-form'
import { useRecoilState } from 'recoil'

import {
  templateTextareaProps,
  templateInputProps,
  modelSentences,
  isSharedProp,
  explainSentences,
  customTemplateSelectProp,
} from './constants'
import { CustomTemplateType } from './type'
import { templateValidationSchema } from './validator'

import { Button, FormControl, Textarea, Switch } from '@/components/molecules'
import { DefaultLayout, ModelSentenceLayout } from '@/components/organisms'
import { user as userState } from '@/contexts/global/atoms/User'
import { usePostCustomTemplate } from '@/hooks/api'

export const CreationTemplate: FC<CustomTemplateType.Props> = memo(({ customTemplates, ...props }) => {
  const { mutate } = usePostCustomTemplate()
  const [user] = useRecoilState(userState)
  const [isSwitchChecked, setIsSwitchChecked] = useState(false)

  useEffect(() => {
    if (customTemplates.length > 0) {
      setIsSwitchChecked(!!customTemplates[0].team)
    }
  }, [customTemplates])

  const {
    register,
    handleSubmit,
    setValue,
    clearErrors,
    formState: { errors, isValid },
    trigger,
  } = useForm<CustomTemplateType.Inputs>({
    defaultValues: {
      isShared: isSwitchChecked,
    },
    mode: 'all',
    resolver: yupResolver(templateValidationSchema),
  })

  const onSubmit = (values: CustomTemplateType.Inputs) => {
    values.id ??= null
    mutate(values)
  }

  const onChange = (value: string) => {
    const valueId = !value ? null : +value
    const template = customTemplates.find((customTemplate) => {
      return customTemplate.id === valueId
    })

    if (!template) {
      return
    }
    const { team, ...rest } = template
    const templateInput: CustomTemplateType.Inputs = { ...rest, isShared: !!team }
    clearErrors()
    ;(Object.keys(templateInput) as (keyof CustomTemplateType.Inputs)[]).forEach(async (key) => {
      if (!valueId && key === 'name') {
        return setValue(key, '')
      }
      if (key === 'isShared') {
        setIsSwitchChecked(templateInput.isShared)
      }
      setValue(key, templateInput[key] ?? '')
    })
  }

  return (
    <DefaultLayout heading="テンプレートを作成・編集する" innerWidth="100%" {...props}>
      <form onSubmit={handleSubmit(onSubmit)}>
        <ModelSentenceLayout modelSentences={modelSentences} explainSentences={explainSentences}>
          <Box gridArea="top_left">
            <Box as="p" mb={12}>
              テンプレートを作成することで、週報を作成する際にテンプレートの内容から週報を書き始めることができます。
              先生から週報の書き方に指定があった際に、そのフォーマットをテンプレートとして登録しておくと便利です。
            </Box>
          </Box>
          <Box gridArea="selectTemplate_left">
            <FormControl label={customTemplateSelectProp.label} isRequired={false} w="100%">
              <Select
                data-cy={`Select_${customTemplateSelectProp.name}`}
                data-testid={`Select_${customTemplateSelectProp.name}`}
                {...(register(customTemplateSelectProp.name),
                {
                  onChange: (e) => {
                    onChange(e.target.value)
                  },
                })}
              >
                {customTemplates.map(({ id, name }, index) => {
                  return (
                    <option value={id ?? ''} key={index}>
                      {name}
                    </option>
                  )
                })}
              </Select>
            </FormControl>
          </Box>
          <Box gridArea="shareSwitch_left">
            {user?.team !== null && (
              <FormControl label={isSharedProp.label} isRequired={false} isInline={true} w="fit-content">
                <Switch
                  {...(register(isSharedProp.name),
                  {
                    onChange: (e) => {
                      onChange(e.target.value)
                    },
                  })}
                  data-testid={`Switch_${isSharedProp.name}`}
                  data-cy={`Switch_${isSharedProp.name}`}
                  isChecked={isSwitchChecked}
                  onChange={() => {
                    const toggle = !isSwitchChecked
                    setValue('isShared', toggle)
                    setIsSwitchChecked(toggle)
                    trigger('isShared')
                  }}
                />
              </FormControl>
            )}
          </Box>
          {templateInputProps.map((inputProp, index) => {
            return (
              <Box gridArea={`${inputProp.name}_left`} key={index}>
                <FormControl label={inputProp.label} error={errors[inputProp.name]}>
                  <Input
                    placeholder={inputProp.placeholder}
                    {...register(inputProp.name)}
                    data-testid={`Input_${inputProp.name}`}
                  />
                </FormControl>
              </Box>
            )
          })}
          {templateTextareaProps.map((textareaProp, index) => {
            return (
              <Box w="100%" gridArea={`${textareaProp.name}_left`} key={index}>
                <FormControl label={textareaProp.label} isRequired={false} error={errors[textareaProp.name]}>
                  <Textarea
                    placeholder={textareaProp.placeholder}
                    {...register(textareaProp.name)}
                    data-testid={`TextArea_${textareaProp.name}`}
                  />
                </FormControl>
              </Box>
            )
          })}

          <Box gridArea="button_left">
            <Center>
              <Button type="submit" isDisabled={!isValid}>
                保存する
              </Button>
            </Center>
          </Box>
        </ModelSentenceLayout>
      </form>
    </DefaultLayout>
  )
})
