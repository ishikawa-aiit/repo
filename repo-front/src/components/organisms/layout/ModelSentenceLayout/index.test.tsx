import { render, screen } from '@testing-library/react'

import { ModelSentenceLayout } from '.'

import { CreationTemplateType } from '@/components/templates/weekly-reports/CreationTemplate/type'

const modelSentence: CreationTemplateType.ModelSentencePropsType = {
  name: 'doneActivity',
  modelSentence: 'これまでの活動と成果の実績',
}

const explainSentence: CreationTemplateType.ExplainSentencePropsType = {
  name: 'selectTemplate',
  explainSentence: 'テンプレートを選択できます',
}

const testText1 = '子コンポーネント1'
const testExplainComponent = <div style={{ gridArea: 'selectTemplate_left' }}>{testText1}</div>
const testText2 = '子コンポーネント2'
const testModelComponent = <div style={{ gridArea: 'doneActivity_left' }}>{testText2}</div>

describe('props:children', () => {
  it('子コンポーネントが正常にレンダリングされている', () => {
    const { container } = render(
      <ModelSentenceLayout modelSentences={[modelSentence]} explainSentences={[explainSentence]}>
        {testExplainComponent}
        {testModelComponent}
      </ModelSentenceLayout>,
    )

    expect(screen.getByText(testText1)).toBeInTheDocument()
    expect(screen.getByText(testText2)).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })
})

describe('ModelSentenceLayout', () => {
  it('テンプレートの特徴文と例文が正常にレンダリングされている', () => {
    const { container } = render(
      <ModelSentenceLayout modelSentences={[modelSentence]} explainSentences={[explainSentence]}>
        {testExplainComponent}
        {testModelComponent}
      </ModelSentenceLayout>,
    )

    expect(screen.getByText(explainSentence.explainSentence)).toBeInTheDocument()
    expect(screen.getByText(modelSentence.modelSentence)).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })
})
