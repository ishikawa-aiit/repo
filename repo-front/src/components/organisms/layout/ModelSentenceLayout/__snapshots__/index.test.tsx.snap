// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModelSentenceLayout テンプレートの特徴文と例文が正常にレンダリングされている 1`] = `
.emotion-0 {
  display: grid;
  grid-template-areas: 'top_left top_right' 'selectTemplate_left selectTemplate_right' 'doneActivity_left doneActivity_right' 'button_left button_right';
  grid-row-gap: 8px;
  grid-column-gap: 8px;
  grid-template-columns: 42rem auto;
}

.emotion-1 {
  grid-area: selectTemplate_right;
  padding-top: 8px;
}

.emotion-2 {
  white-space: pre-wrap;
  color: neutral.500;
  font-weight: normal;
}

.emotion-3 {
  grid-area: doneActivity_right;
  padding-top: 8px;
}

.emotion-4 {
  color: neutral.500;
  font-weight: normal;
  font-size: inherit;
}

.emotion-6 {
  grid-area: button_right;
}

<div>
  <div
    class="emotion-0"
  >
    <div
      style="grid-area: selectTemplate_left;"
    >
      子コンポーネント1
    </div>
    <div
      style="grid-area: doneActivity_left;"
    >
      子コンポーネント2
    </div>
    <div
      class="emotion-1"
    >
      <p
        class="chakra-text emotion-2"
      >
        テンプレートを選択できます
      </p>
    </div>
    <div
      class="emotion-3"
    >
      <h2
        class="chakra-heading emotion-4"
      >
        例)
      </h2>
      <p
        class="chakra-text emotion-2"
      >
        これまでの活動と成果の実績
      </p>
    </div>
    <div
      class="emotion-6"
    />
  </div>
</div>
`;

exports[`props:children 子コンポーネントが正常にレンダリングされている 1`] = `
.emotion-0 {
  display: grid;
  grid-template-areas: 'top_left top_right' 'selectTemplate_left selectTemplate_right' 'doneActivity_left doneActivity_right' 'button_left button_right';
  grid-row-gap: 8px;
  grid-column-gap: 8px;
  grid-template-columns: 42rem auto;
}

.emotion-1 {
  grid-area: selectTemplate_right;
  padding-top: 8px;
}

.emotion-2 {
  white-space: pre-wrap;
  color: neutral.500;
  font-weight: normal;
}

.emotion-3 {
  grid-area: doneActivity_right;
  padding-top: 8px;
}

.emotion-4 {
  color: neutral.500;
  font-weight: normal;
  font-size: inherit;
}

.emotion-6 {
  grid-area: button_right;
}

<div>
  <div
    class="emotion-0"
  >
    <div
      style="grid-area: selectTemplate_left;"
    >
      子コンポーネント1
    </div>
    <div
      style="grid-area: doneActivity_left;"
    >
      子コンポーネント2
    </div>
    <div
      class="emotion-1"
    >
      <p
        class="chakra-text emotion-2"
      >
        テンプレートを選択できます
      </p>
    </div>
    <div
      class="emotion-3"
    >
      <h2
        class="chakra-heading emotion-4"
      >
        例)
      </h2>
      <p
        class="chakra-text emotion-2"
      >
        これまでの活動と成果の実績
      </p>
    </div>
    <div
      class="emotion-6"
    />
  </div>
</div>
`;
